### ISSUE TITLE
Add support for S3 server side encryption using user keys ( known as SSE-C )

### ISSUE THREAD

**Issue by robromano** [boto/boto#2334]
See http://aws.amazon.com/about-aws/whats-new/2014/06/12/amazon-s3-now-supports-server-side-encryption-with-customer-provided-keys-sse-c/

This ticket is to enhance boto to support this.

**Comment by sanislav** [comment]
@danielgtaylor Your example doesn't work for me. Using the password `01testKeyToSSEC!` the 2 values I get are:

encryption_key = `wbTfEmLFcHNwm+WR03SY96khq4eS/yaadFGjRww+XVg=`
and
encryption_key_md5 = `RHXzDK4Kt2+zfpfgnCDS2Q==`

and not `MAAxAHQAZQBzAHQASwBlAHkAVABvAFMAUwBFAEMAIQA=` and `fUgCZDDh6bfEMuP2bN38mg==`

@nbermudezs did you have any luck with this?

**Comment by nbermudezs** [comment]
Sadly, no. Since my requirements didn't force me to use SSE-C I gave up after a while and ended up using SSE with Managed Encryption Keys instead.

**Comment by sanislav** [comment]
@nbermudezs @danfairs I figured it out. First of all the test uses `01testKeyToSSEC!` as password.

Then the methods to generate the 2 keys are:
`
password = '32characterslongpassphraseneeded'
encryption_key = base64.b64encode(password)
encryption_key_md5 = base64.b64encode(hashlib.md5(password).digest())
`

Thank you.

**Comment by nbermudezs** [comment]
But then doing this
```python
password = '01testKeyToSSEC!'
encryption_key = base64.b64encode(password)
encryption_key_md5 = base64.b64encode(hashlib.md5(password).digest())
```

Gives me 
```python
encryption_key = 'MDF0ZXN0S2V5VG9TU0VDIQ=='
encryption_key_md5 = 'RHXzDK4Kt2+zfpfgnCDS2Q=='
```
Which are not the values sent in the [headers](https://github.com/boto/boto/blob/c8c625ad09246e1a835f4a05174e8ed1495cbb00/tests/integration/s3/test_key.py#L448-L455) for the test file. And I know for fact that those values work but who knows how they generated them :confused: 

Anyways, I'm glad that worked for you. Thanks for sharing it.

**Comment by mmmonk** [comment]
ok, so if you take:

    base64.b64decode('MAAxAHQAZQBzAHQASwBlAHkAVABvAFMAUwBFAEMAIQA=').encode('hex')
    '3000310074006500730074004b006500790054006f0053005300450043002100'

and the above matches this:

    "".join([a+'\x00' for a in '01testKeyToSSEC!']).encode('hex')
    '3000310074006500730074004b006500790054006f0053005300450043002100'

and then:
    
    base64.b64decode('fUgCZDDh6bfEMuP2bN38mg==').encode('hex')
    '7d48026430e1e9b7c432e3f66cddfc9a'

so it seems to get that we need to do this:

    hashlib.md5("".join([a+'\x00' for a in '01testKeyToSSEC!'])).hexdigest()
    '7d48026430e1e9b7c432e3f66cddfc9a'

so to summarise the correct syntax should be something along the lines of:

    password = '01testKeyToSSEC!'
    encryption_key = base64.b64encode("".join([a+'\x00' for a in password]))
    encryption_key_md5 = base64.b64encode(hashlib.md5("".join([a+'\x00' for a in password])).digest())

if anybody still needs this.

**Comment by Darwesh27** [comment]
Hey Guys. I am using python 3.4.3. And I am using BotoS3Storage as the default storage for my app.
And for encryption purposes I added the key in headers as described in [this](http://docs.aws.amazon.com/AmazonS3/latest/dev/ServerSideEncryptionCustomerKeys.html) documentation. And here is how I do it.

> #settings.py
> AWS_S3_ENCRYPTION = True
> password = '32characterslongpassphraseneeded'
> encryption_key = base64.b64encode(password)
> encryption_key_md5 = base64.b64encode(hashlib.md5(password).digest())
> 
> AWS_HEADERS = {
>         'x-amz-server-side-encryption-customer-algorithm': 'AES256',
>     'x-amz-server-side-encryption-customer-key': encryption_key,
>     'x-amz-server-side-encryption-customer-key-MD5': encryption_key_md5,
>     'Cache-Control': 'max-age=86400',
> }
> 
But I got the error. "TypeError: 'str' does not support the buffer interface" for password. So I changed the line to. 

> password = '32characterslongpassphraseneeded'.encode('utf-8')

Now I get the following error. 

> [Errno 104] Connection reset by peer

The file uploads work fine If I don't specify any headers. There seems to be some problem with the headers. 

Can I get some help.
